storage/remote/cache.go

```diff
@@ -0,0 +1,55 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"sync"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+type labelsString string
+
+// seriesCache stores the position of a series entry in the shard's queue.
+type seriesCache struct {
+	mux   sync.RWMutex
+	size  int
+	cache map[labelsString]int
```

Comment:
It would be really nice if we could use a ref for this instead of the labelString. Have you benchmarked this at all, I could imagine the lookups being fairly expensive, but could be wrong.